core/drivers/crypto/crypto_api/acipher/rsassa.c

```diff
@@ -0,0 +1,918 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2020 NXP
+ *
+ * Brief   RSA Signature Software common implementation.
+ *         Functions preparing and/or verifying the signature
+ *         encoded string.
+ *
+ *         PKCS #1 v2.1: RSA Cryptography Standard
+ *         https://www.ietf.org/rfc/rfc3447.txt
+ */
+#include <crypto/crypto.h>
+#include <drvcrypt.h>
+#include <drvcrypt_asn1_oid.h>
+#include <drvcrypt_math.h>
+#include <malloc.h>
+#include <string.h>
+#include <tee_api_defines_extensions.h>
+#include <tee/tee_cryp_utl.h>
+#include <utee_defines.h>
+#include <util.h>
+
+#include "local.h"
+
+/*
+ * PKCS#1 V1.5 - Encode the message in Distinguished Encoding Rules
+ * (DER) format.
+ * Refer to EMSA-PKCS1-v1_5 chapter of the PKCS#1 v2.1
+ *
+ * @ssa_data  RSA data to encode
+ * @EM        [out] Encoded Message
+ */
+static TEE_Result emsa_pkcs1_v1_5_encode(struct drvcrypt_rsa_ssa *ssa_data,
+					 struct cryptobuf *EM)
+{
+	const struct drvcrypt_oid *hash_oid = NULL;
+	size_t ps_size = 0;
+	uint8_t *buf = NULL;
+
+	hash_oid =
+		&drvcrypt_hash_oid[TEE_ALG_GET_MAIN_ALG(ssa_data->hash_algo)];
+
+	/*
+	 * Calculate the PS size
+	 *  EM Size (modulus size) - 3 bytes - DigestInfo DER format size
+	 */
+	ps_size = ssa_data->key.n_size - 3;
+	ps_size -= ssa_data->digest_size;
+	ps_size -= 10 + hash_oid->asn1_length;
+
+	CRYPTO_TRACE("PS size = %zu (n %zu)", ps_size, ssa_data->key.n_size);
+
+	/*
+	 * EM = 0x00 || 0x01 || PS || 0x00 || T
+	 *
+	 * where T represent the message DigestInfo in DER:
+	 *    DigestInfo ::= SEQUENCE {
+	 *                digestAlgorithm AlgorithmIdentifier,
+	 *                digest OCTET STRING
+	 *                }
+	 *
+	 * T  Length = digest length + oid length
+	 * EM Length = T Length + 11 + PS Length
+	 */
+	buf = EM->data;
+
+	/* Set the EM first byte to 0x00 */
+	*buf++ = 0x00;
+
+	/* Set the EM second byte to 0x01 */
+	*buf++ = 0x01;
+
+	/* Fill PS with 0xFF */
+	memset(buf, UINT8_MAX, ps_size);
+	buf += ps_size;
+
+	/* Set the Byte after PS to 0x00 */
+	*buf++ = 0x00;
+
+	/*
+	 * Create the DigestInfo DER Sequence
+	 *
+	 *  DigestInfo ::= SEQUENCE {
+	 *                digestAlgorithm AlgorithmIdentifier,
+	 *                digest OCTET STRING
+	 *                }
+	 *
+	 */
+	/* SEQUENCE { */
+	*buf++ = ASN1_SEQUENCE | ASN1_CONSTRUCTED;
+	*buf++ =
+		(uint8_t)(0x08 + hash_oid->asn1_length + ssa_data->digest_size);
+
+	/* digestAlgorithm AlgorithmIdentifier */
+	*buf++ = ASN1_SEQUENCE | ASN1_CONSTRUCTED;
+	*buf++ = (uint8_t)(0x04 + hash_oid->asn1_length);
```

Comment:
It looks like we could do without this cast.

Reply:
Ok